pkg/apis/build/v1alpha1/parameter.go

```diff
@@ -4,9 +4,51 @@
 
 package v1alpha1
 
+type ObjectKeyRef struct {
+
+	// Name of the object
+	// +required
+	Name string `json:"name"`
+
+	// Key inside the object
+	// +required
+	Key string `json:"key"`
+
+	// An optional format to add pre- or suffix to the object value. For example 'KEY=${SECRET_VALUE}' or 'KEY=${CONFIGMAP_VALUE}' depending on the context.
+	// +optional
+	Format *string `json:"format"`
+}
+
+// The value type contains the properties for a value, this allows for an
+// easy extension in the future to support more kinds
+type SingleValue struct {
+
+	// The value of the parameter
+	// +optional
+	Value *string `json:"value"`
+
+	// The ConfigMap value of the parameter
+	// +optional
+	ConfigMapValue *ObjectKeyRef `json:"configMapValue"`
+
+	// The secret value of the parameter
+	// +optional
+	SecretValue *ObjectKeyRef `json:"secretValue"`
+}
+
 // ParamValue is a key/value that populates a strategy parameter
 // used in the execution of the strategy steps
 type ParamValue struct {
-	Name  string `json:"name"`
-	Value string `json:"value"`
```

Comment:
Going back and looking at the SHIP, it was not clearly evident to me that the `Value` field would disappear, in favor the new fields.

But even if that was understood, and yes we are still an alpha API, I would be inclined to be "nicer" in transitioning existing users to the new API.

The field could be left in but marked deprecated as part of this change going out for v0.8.0

This feels analogous to how we handled the implicit tekton workspace emptydir volumes with https://github.com/shipwright-io/build/pull/941

I think we could minimize its short term existence in the implementation by having the build/buildrun reconcilers up top look for instances of it being set and copying it to `*SingleValue`.

Then in v0.9.0 we remove it.

WDYT @SaschaSchwarze0 ?

bump @adambkaplan - given you were the original advocate for what we did with https://github.com/shipwright-io/build/pull/941 and making progress on SHIP-22

Reply:
Hi @gabemontero, the `value` field is not going away in the serialized (yaml/json) format. In `ParamValue`, I am inlining `SingleValue` which brings back the `value` field. So, we are here breaking "only" the Go API (like we also do in Shahul's PR where we fix the handling of optional fields). But the "YAML API" is not broken.

The alternative of not inlining is to have `configMapValue`, `secretValue`, and `value` copied into `ParamValue`. That would be okay for me as well. But, I would still need to break the Go API. Previously value was not optional and no pointer. Now it is both because it is just one of the three possible fields that can be set to define a value (and in case of an array, none of them would be set).